clients/src/main/java/org/apache/kafka/clients/Metadata.java

```diff
@@ -146,17 +148,32 @@ public synchronized boolean updateRequested() {
     }
 
     /**
+     * If any non-retriable authentication exceptions were encountered during
+     * metadata update, clear and throw the exception.
+     */
+    public synchronized void maybeHandleNonRetriableAuthenticationExceptions() {
+        if (nonRetriableAuthenticationException != null) {
+            AuthenticationException exception = nonRetriableAuthenticationException;
+            this.needUpdate = false;
```

Comment:
Curious what the reason for this is.

Reply:
It is to avoid unnecessary connections being made to update metadata e.g. from `Sender` thread which will continue to retry if `needUpdate` is set. The flag will be set again if the application performs another operation (e.g. another send), but we are avoiding automatic retry.

Reply:
I see. My concern is that we might lose track of the fact that we still need the metadata update. For example, we request a metadata when the user calls `consumer.subscribe(Pattern)`. If we then raise an `AuthenticationException` from `poll()` and the user ignores it, then we might not bother attempting another metadata update until the next refresh time. Maybe that is an acceptable tradeoff to avoid having the background threads retrying needlessly until shutdown? At a minimum, it would be useful to have a comment here explaining the purpose.

Reply:
Hmm.. In the consumer, we only fetch metadata if we know we need to, and that is sometimes conveyed only through the `needUpdate` flag (as in the case I mentioned).

One additional note: this flag alone is not sufficient to prevent metadata retries if the metadata max age has expired (which will be the case on initialization).

Reply:
@hachikuji  I may have missed something, but in the consumer case, we have no coordinator or leader information since we could not fetch metadata. So we will have to refresh on the next `poll()`. Both `ensureCoordinatorReady` and the `Fetcher` request updates.

On the second note: I was thinking that we would clear the topics on authentication exception in producer to avoid automatic refresh on expiry, but I had forgotten about that. Perhaps we should? On the consumer, I was thinking this is not an issue since you have to explicitly `poll()` to trigger an update - have I missed something?

Reply:
@rajinisivaram Yeah, I think we are handling the common (initialization) cases in this patch, just a bit discomforting to leave some of the other cases a little nebulously handled. 

In this particular case, it doesn't really seem worth trying to prevent additional metadata retries since the NetworkClient will still enforce a backoff and we expect the user to just close the client anyway. I would probably leave `needUpdate` untouched as we do currently on failure.

Reply:
@hachikuji OK, I am fine with that. @vahidhashemian Can you remove the `needUpdate` change here? Thanks.

Reply:
Sure. Thanks.